dialer_test.go

```diff
@@ -550,17 +582,92 @@ func TestDialerRemovesInvalidInstancesFromCache(t *testing.T) {
 	}
 }
 
-type spyConnectionInfoCache struct {
-	connectInfoError error
+func TestDialRefreshesExpiredCertificates(t *testing.T) {
+	d, err := NewDialer(context.Background(),
+		WithTokenSource(mock.EmptyTokenSource{}),
+	)
+	if err != nil {
+		t.Fatalf("expected NewDialer to succeed, but got error: %v", err)
+	}
+
+	sentinel := errors.New("connect info failed")
+	icn := "project:region:instance"
+	cn, _ := cloudsql.ParseConnName(icn)
+	spy := &spyConnectionInfoCache{
+		connectInfoCalls: []struct {
+			tls *tls.Config
+			err error
+		}{
+			// First call returns expired certificate
+			{
+				tls: &tls.Config{
+					Certificates: []tls.Certificate{{
+						Leaf: &x509.Certificate{
+							// Certificate expired 10 hours ago.
+							NotAfter: time.Now().Add(-10 * time.Hour),
+						},
+					}},
+				},
+			},
+			// Second call errors to validate error path
```

Comment:
Would it be clearer if the error path was in a separate test case?

Reply:
The implementation makes two calls to ConnectInfo. To test this case, we need to ensure the first call returns an expired cert. For the second call, we return an error to test the extra error path as a convenience. Otherwise, we'd have to do a lot more wiring up of TLS config, server TLS, etc, which might be worth doing, but probably not in this PR.